core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -82,48 +100,157 @@ public void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardR
+            // we're not reusing an existing in-sync allocation id to initialize a primary, which means that we're either force-allocating
+            // an empty or a stale primary (see AllocateEmptyPrimaryAllocationCommand or AllocateStalePrimaryAllocationCommand).
+            boolean emptyPrimary = updates.initializedPrimary.unassignedInfo().getReason() == UnassignedInfo.Reason.INDEX_CREATED;
+            assert updates.addedAllocationIds.isEmpty() : (emptyPrimary ? "empty" : "stale") +
+                " primary is not force-initialized in same allocation round where shards are started";
+
+            if (indexMetaDataBuilder == null) {
+                indexMetaDataBuilder = IndexMetaData.builder(oldIndexMetaData);
+            }
+            if (emptyPrimary) {
+                // forcing an empty primary resets the in-sync allocations to the empty set (ShardRouting.allocatedPostIndexCreate)
+                indexMetaDataBuilder.putInSyncAllocationIds(shardId.id(), Collections.emptySet());
+            } else {
+                // forcing a stale primary resets the in-sync allocations to the singleton set with the stale id
+                indexMetaDataBuilder.putInSyncAllocationIds(shardId.id(),
+                    Collections.singleton(updates.initializedPrimary.allocationId().getId()));
+            }
+        } else {
+            // standard path for updating in-sync ids
+            Set<String> inSyncAllocationIds = new HashSet<>(oldInSyncAllocationIds);
+            inSyncAllocationIds.addAll(updates.addedAllocationIds);
+            inSyncAllocationIds.removeAll(updates.removedAllocationIds);
+
+            // Prevent set of inSyncAllocationIds to grow unboundedly. This can happen for example if we don't write to a primary
+            // but repeatedly shut down nodes that have active replicas.
+            // We use number_of_replicas + 1 (= possible active shard copies) to bound the inSyncAllocationIds set
+            int upperBound = oldIndexMetaData.getNumberOfReplicas() + 1; // +1 for the primary
+            int overflow = inSyncAllocationIds.size() - upperBound;
+            if (overflow > 0) {
+                // trim entries that have no corresponding shard routing in the cluster state (i.e. trim unavailable copies)
+                Set<String> assignedAllocations = newRoutingTable.shardRoutingTable(shardId).assignedShards().stream()
+                    .map(shr -> shr.allocationId().getId()).collect(Collectors.toSet());
+                Set<String> candidatesToRemove = Sets.difference(inSyncAllocationIds, assignedAllocations);
+                for (Iterator<String> it = candidatesToRemove.iterator(); it.hasNext() && overflow > 0; ) {
+                    inSyncAllocationIds.remove(it.next());
+                    overflow--;
+                }
+                assert overflow == 0 :
+                    "not enough candidates to remove, candidates: " + candidatesToRemove + " overflow: " + overflow;
```

Comment:
This assert message won't be too helpful because it won't show the original overflow, but it will show the original candidates to remove, for example, you could end up with:

```
not enough candidates to remove, candidates: a, b, c, d, e, f, g, h, i, j, k overflow: 1
```

perhaps it should show the ones that couldn't be removed that could still be trimmed?